src/lib/dovetail_audio.ts

```diff
@@ -0,0 +1,221 @@
+      value: audio,
+      writable: false
+    });
+    if (extras) {
+      for (let property in extras) {
+        if (extras.hasOwnProperty(property)) {
+          event[property] = extras[property];
+        }
+      }
+    }
+    return event;
+  }
+}
+
+export class DovetailAudio extends ExtendableAudio {
+  private arrangement: Arrangement = {entries: []};
+  private index: number = 0;
+
+  private _dovetailLoading = false;
+
+  private adzerkFetcher = new AdzerkFetcher();
+  private dovetailFetcher = new DovetailFetcher();
+
+  private currentPromise: Promise<any>;
+  private currentAdzerkPromise: Promise<AdzerkResponse>;
+  private resumeOnLoad = false;
+
+  constructor(url: string) {
+    super(url);
+    this._audio.addEventListener(DURATION_CHANGE, this.listenerOnDurationChange.bind(this));
+    this._audio.addEventListener(ENDED, this.listenerOnEnded.bind(this));
+    this._audio.addEventListener(TIME_UPDATE, this.listenerOnTimeUpdate.bind(this));
+    this.finishConstructor();
+  }
+
+  play() {
+    if (this._dovetailLoading) {
+      this.resumeOnLoad = true;
+    } else {
+      this._audio.play();
+    }
+  }
+
+  pause() {
+    if (this._dovetailLoading) {
+      this.resumeOnLoad = false;
+    } else {
+      this._audio.pause();
+    }
+  }
+
+  get duration() {
+    if (typeof this.arrangement.duration === 'undefined') {
+      this.arrangement.duration = this.arrangement.entries.reduce(sumDuration, 0);
+    }
+    return this.arrangement.duration;
+  }
+
+  get currentTime() {
+    return this._audio.currentTime +
+      this.arrangement.entries.slice(0, this.index).reduce(sumDuration, 0);
+  }
+
+  set currentTime(position: number) {
+    if (this.duration >= position) {
+      let soFar = 0, paused = this.paused;
+      for (let i = 0; i < this.arrangement.entries.length; i++) {
+        let duration = this.arrangement.entries[i].duration;
+        if (soFar + duration > position) {
+          this.skipToFile(i);
+          this._audio.currentTime = position - soFar;
+          if (!paused) { this.play(); }
+          return;
+        } else {
+          soFar += duration;
+        }
+      }
+    }
+  }
+
+  set src(url: string) {
+    this._dovetailLoading = true;
+    let promise = this.currentPromise = this.dovetailFetcher.fetch(url).then(
+      result => {
+        if (this.currentPromise == promise) {
+          let data: AllUnion  = [result, this.getArrangement(result.request)];
+          return Promise.all<any>(data);
+        }
+      }
+    ).then(([dovetailResponse, adzerkResponse]: AllResult) =>  {
+      return this.calculateArrangement(dovetailResponse.arrangement, adzerkResponse);
+    }).catch(error => {
+      return [{
+        audioUrl: url,
+        duration: 0,
+        id: 'fallback',
+        type: 'fallback'
+      }];
+    }).then(arrangement => {
+      if (this.currentPromise == promise) {
+        this._dovetailLoading = false;
+        this.setSegments(arrangement);
+      }
+    });
+  }
```

Comment:
This is where the real magic happens, when you set the src url, we attempt to load the dovetail design doc and make a request to adzerk. If any part of that fails (adblockers or a non-dovetail url), we fall back to a `fallback` arrangement which just uses the original URL.